### DISCUSSION TITLE
Moose Installation - HPC

### DISCUSSION THREAD

**Discussion by normjona** [idaholab/moose#22659] [Q&A Getting Started] [answered]
Hi, 

I'm installing moose and BISON on an HPC cluster and have been running into what appears to be a libmesh related error.  

After successfully running the update and rebuild PETSc script, I run run the update and rebuild libmesh script.  I get this error
```
---------------------------------------------
-- Done configuring core library features ---
---------------------------------------------
---------------------------------------------
----- Configuring for optional packages -----
---------------------------------------------
checking for built-in XDR support... no
checking for XDR support in /usr/include/tirpc... no
configure: error: *** XDR was not found, but --enable-xdr-required was specified.
Running make -j 1...
make: *** No targets specified and no makefile found.  Stop.
```
After this when I try to run the make command, I get the following errors. 
```
make -j 4
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Traceback (most recent call last):
  File "/nfs/home/jnorma18/projects/moose/framework/scripts/get_repo_revision.py", line 19, in <module>
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
Checking if header needs updating: /nfs/home/jnorma18/projects/moose/framework/include/base/MooseRevision.h...
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Compiling C++ (in opt mode) /nfs/home/jnorma18/projects/moose/framework/build/unity_src/src_Unity.C...
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/libtool: No such file or directory
Compiling C++ (in opt mode) /nfs/home/jnorma18/projects/moose/framework/build/unity_src/actions_Unity.C...
make: *** [/nfs/home/jnorma18/projects/moose/framework/build.mk:144: /nfs/home/jnorma18/projects/moose/framework/build/unity_src/src_Unity..opt.lo]
 Error 127
make: *** Waiting for unfinished jobs....
/bin/sh: /nfs/home/jnorma18/projects/moose/libmesh/installed/libtool: No such file or directory
make: *** [/nfs/home/jnorma18/projects/moose/framework/build.mk:145: /nfs/home/jnorma18/projects/moose/framework/build/unity_src/actions_Unity..opt
.lo] Error 127

```
My libmesh directory isn't including the "install" directory.  It seems libmesh is failing to build correctly?  I'd appreciate any help on this.

Thanks,
Jonathan Norman

**Comment by GiudGiud** [comment, marked as answer]
hello

Libmesh did not build properly.
The moose error you got afterwards is 'libmesh not installed' basically
You need to install this library on your cluster `libtirpc-devel` then try to install libmesh again

Guillaume

**Reply by jessecarterMOOSE** [reply]
@GiudGiud Got this same error today when testing out a new HPC cluster. I noticed I have libtirpc.so.3 installed in /lib64, but libmesh config  still doesn't find XDR.

**Reply by GiudGiud** [reply]
what does it report under:
```
---------------------------------------------
----- Configuring for optional packages -----
---------------------------------------------
```
in the log

**Reply by jessecarterMOOSE** [reply]
```
checking for built-in XDR support... no
checking for XDR support in /usr/include/tirpc... no
configure: error: *** XDR was not found, but --enable-xdr-required was specified.
```

Just checked and there is no /usr/include/tirpc directory. I'm not an admin for the cluster either :cry:

**Reply by GiudGiud** [reply]
can you get them to install `libtirpc-devel` still?

**Reply by jessecarterMOOSE** [reply]
I'll ask

**Reply by ajacquey** [reply]
Hi @GiudGiud ,

I'm reopening this thread as I encountered the same issue on a HPC cluster.
I contacted the administrator of the cluster to have `libtirpc-devel` installed but I was told the library is already installed but not in `/usr/inlude/tirpc` and that the compiler (loaded by module) is correctly linked to the proper location where the library is installed.
The administrator suggests that `libmesh` is only looking into `/usr/include` or using `/usr/bin/gcc` instead of the default compiler (from the module).  I doubt the second option as from the log, `libmesh` is using the correct `gcc` compiler (and not `/usr/bin/gcc`)

Do you know anything about it? Is it possible that `libmesh` is only looking at `/usr/include/` for the `libtirpc` library?

**Reply by GiudGiud** [reply]
not sure. Can you play with the environment flags to include `/usr/bin/gcc` at link time?

**Reply by GiudGiud** [reply]
if that does not work we ll ask Roy where libmesh looks

**Reply by jessecarterMOOSE** [reply]
@GiudGiud My HPC admins installed libtirpc but as a module (`module load libtirpc`) so the header files go their own installation directory somewhere else and not /usr/include. How can I tell libmesh to look in this other directory?

**Reply by ajacquey** [reply]
@jessecarterMOOSE I managed to solve that by declaring the following variable before running the `./update_and_rebuild_libmesh.sh` :
```
export CPPFLAGS=-I/path/to/your/libtirpc/installation/
```

**Reply by jessecarterMOOSE** [reply]
@ajacquey did you get an indication from the libmesh configure output that it was honoring that environment variable? I tried what you say and I still see 
```
checking for built-in XDR support... no
checking for XDR support in /usr/include/tirpc... no
configure: error ...
```

So it doesn't seem to be checking the libtirpc installation directory? I've tried:

```
CPPFLAGS=-I/path/to/libtirpc
CPPFLAGS=-I/path/to/libtirpc/include
CPPFLAGS=-I/path/to/libtirpc/include/tirpc
```

and none work :(

**Reply by GiudGiud** [reply]
@milljm

**Reply by ajacquey** [reply]
@jessecarterMOOSE Yes my libmesh log indicates that it finds tirpc.
In my case, I used:
```
export CPPFLAGS=-I/$EBROOTGENTOO/include/tirpc/
```
where `EBROOTGENTOO` is an environmental variable on the cluster. In the folder linked above, I have:
```
netconfig.h
rpc/
rpcsvc/
```

**Reply by jessecarterMOOSE** [reply]
@ajacquey Did you see anything in the libmesh configure output that mentions your libtirpc install directory? Anything related to XDR?

**Reply by milljm** [reply]
@roystgnr Is there an environment variable that controls libMesh where to look for `tirpc` headers?

**Reply by GiudGiud** [reply]
not specific to rpc but there is `libmesh_INCLUDE` right?

**Reply by roystgnr** [reply]
Currently it's hard-coded, to first look for `rpc/rpc.h` and `rpc/xdr.h` in the default search paths, then to add `-I/usr/include/tirpc` to `$CPPFLAGS` and `-ltirpc` to `$LIBS` and try again.  Something like `configure CPPFLAGS=-I/$EBROOTGENTOO/include/tirpc/ LIBS='-L$EBROOTGENTOO/lib/ -Wl,-rpath,$EBROOTGENTOO/lib/ -ltirpc'` might be sufficient to get that install?  I've done something like that before to manually combine a gcc-built-mpich with a clang++ compiler.

**Reply by jessecarterMOOSE** [reply]
@roystgnr you're a genius! Not sure how that worked, but it did.

**Reply by roystgnr** [reply]
There's a fine line between genius and madness.  By the time you've figured out that you need both `-L` and `-Wl,-rpath` to make compile-time and run-time link paths clear, you're generally on the wrong side of the line...

Are people really having to compile their own RPC libraries these days?  Should we be adding a `TIRPC_DIR=$foo` configure/environment option?

**Reply by jessecarterMOOSE** [reply]
I was told that RPC was removed in whatever version of glibc was installed on our cluster and it's a standalone package now.

**Reply by jessecarterMOOSE** [reply]
This issue seems to be resolved (for me) after the latest libmesh update.

**Reply by jessecarterMOOSE** [reply]
libmesh configure just showed "checking for built-in XDR support... yes", so I'm not sure how it's finding the libraries, but at least it does

**Reply by echo1115** [reply]
Hi @GiudGiud 
I'm reopening this thread since I encountered the same issue:
```
checking for built-in XDR support... no
checking for XDR support in /usr/include/tirpc... no
configure: error: *** XDR was not found, but --enable-xdr-required was specified.
```
I asked the cluster administrator to install `libtirpc-devel`, but he said that in order to ensure that it does not affect other software in our cluster, it is stipulated that `libtirpc-devel` cannot be installed in the default directory, but must be installed in a specified directory.
I tried installing it into a specified directory  under my account, but failed:
```
rpm -ivh --prefix=/home/dcq/depency_my/libtirpc-devel/ libtirpc-devel-1.1.4-5.el8.x86_64.rpm 
error: package libtirpc-devel is not relocatable
```

I noticed in the previous discussion that @ajacquey said his administrator installed libtirpc-devel to a different directory instead of `/usr/inlude/tirpc` :
> I'm reopening this thread as I encountered the same issue on a HPC cluster. I contacted the administrator of the cluster to have `libtirpc-devel` installed but I was told the library is already installed but not in `/usr/inlude/tirpc` and that the compiler (loaded by module) is correctly linked to the proper location where the library is installed. The administrator suggests that `libmesh` is only looking into `/usr/include` or using `/usr/bin/gcc` instead of the default compiler (from the module). I doubt the second option as from the log, `libmesh` is using the correct `gcc` compiler (and not `/usr/bin/gcc`)
> 


  I would like to know how to achieve this?

**Reply by roystgnr** [reply]
>  in order to ensure that it does not affect other software in our cluster, it is stipulated that libtirpc-devel cannot be installed in the default directory, but must be installed in a specified directory.

"We didn't install a basic system library, so some of our users probably had to install their own version, so to avoid conflicts we're not going to install a basic system library", huh?  ;-)  My condolences.

You might be stuck installing from source.  Looks like the official git repo is at https://git.linux-nfs.org/?p=steved/libtirpc.git;a=summary and 1.3.4 is the latest release version.

**Reply by echo1115** [reply]
Thanks for your reply @roystgnr 
Actually, I need to offline install MOOSE on a RHEL 8.4 cluster. When encountering the configure error for libmesh, I run this:
```
[dcq@localhost ~]$ rpm -q libtirpc
libtirpc-1.1.4-4.el8.x86_64
```
indicating that libtirpc is already packaged via rpm in the system, but` libtirpc-devel` is missing. Then I downloaded the libtirpc-devel rpm package suitable for CentOS 8 from this website: [link](https://www.rpmfind.net/linux/rpm2html/search.php?query=libtirpc-devel&submit=Search+...&system=&arch=), but I'm unable to install the 'libtirpc-devel-1.1.4-5.el8.x86_64.rpm' to the directory I want.

Regarding the link you provided:

> You might be stuck installing from source. Looks like the official git repo is at https://git.linux-nfs.org/?p=steved/libtirpc.git;a=summary and 1.3.4 is the latest release version.

I assume it's for installing libtirpc rather than libtirpc-devel.

**Reply by roystgnr** [reply]
> indicating that libtirpc is already packaged via rpm in the system, but libtirpc-devel is missing

Wait, your sysadmin said he didn't want to install **the headers used to compile against already-installed libraries** in order to avoid conflicts with other software?  That's ... a philosophy, I guess?  Pity the poor developers like you who have to install their own TIRPC to get it work and then *also* have to make sure it's either ABI-compatible with or not conflicting with the system version.

The link I provided is for all of TIRPC, which is both tirpc and tirpc-devel.  RedHat breaks these packages into multiple different rpms, so that end users can install the parts (e.g. libraries, in the tirpc .rpm) needed to run already-compiled programs without having to also install the parts (e.g. header files, in the tirpc-devel .rpm) needed to compile new programs.  Ideally you'd simply build all of TIRPC yourself and link against it... but if the TIRPC rpm is already installed you'll need to either keep its (system!) library directory out of your linker path, or you'll have to get the version matching the tirpc version that's installed, cross your fingers hoping there aren't any ABI-incompatible patches on that rpm's version, and use the headers you download to link to the libraries already installed.

**Reply by echo1115** [reply]
@roystgnr Thank you for detailed explanation and sharing your guidance on this matter. 
Now I've decided to proceed with installing TIRPC from the source code provided in the link you shared. During the process, I've encountered many missing packages in my system, such as autoconf, automake, libtool, and others. Similarly, I'll need to install them from source code, and to install them, I need to install additional tools or modules. It feels like a never-ending process, and I'm unsure where it ends. However, I hope that this will eventually lead to successfully installing TIRPC, enabling me to proceed with compiling and rebuilding libmesh.

**Reply by milljm** [reply]
For these types of dependencies, I like to use a clean Conda environment to satisfy.

1. Install Conda
2. Create a pristine environment, purposefully only containing what is necessary (whats missing from the System):
```bash
conda create -n libmesh-dependencies libxt-devel-cos7-x86_64=1.1.5
conda activate libmesh-dependencies
```
And then try to run `./update_and_rebuild_libmesh.sh` again. The above should take care of `libtirpc`.

Upon the next error (if there is any), install that one single requirement while in the `libmesh-dependencies` Conda environment, and try again. Repeat.

If memory serves, you might run into the next set of missing 'things'. Do not install unless necessary!
```pre
autoconf
automake
m4
libtool    # I would hope not
```

**Reply by echo1115** [reply]
@milljm Thank you for your suggestion! It is also a good idea
